File: src/waypalm-site/src/main/java/waypalm/site/validation/ReCaptchaValidator.java
package waypalm.site.validation;

import org.springframework.beans.BeanWrapper;
import org.springframework.beans.PropertyAccessorFactory;
import waypalm.site.services.RecaptchaValidator;

import javax.inject.Inject;
import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;

public class ReCaptchaValidator implements ConstraintValidator<ReCaptcha, Object> {
    private String challengeField;
    private String responseField;

    @Inject
    private RecaptchaValidator captchaValidator;

    @Override
    public void initialize(ReCaptcha ann) {
        challengeField = ann.challengeField();
        responseField = ann.responseField();
    }

    @Override
    public boolean isValid(Object value, ConstraintValidatorContext context) {
        BeanWrapper beanWrapper = PropertyAccessorFactory.forBeanPropertyAccess(value);
        String challenge = (String) beanWrapper.getPropertyValue(challengeField);
        String response = (String) beanWrapper.getPropertyValue(responseField);
        return captchaValidator.validate(challenge, response);
    }
}

File: src/waypalm-site/src/main/java/waypalm/site/validation/ReCaptchaValidator.kt
package waypalm.site.validation

import javax.validation.ConstraintValidator
import javax.inject.Inject
import waypalm.site.services.RecaptchaValidator
import javax.validation.ConstraintValidatorContext
import javax.validation.Constraint


Documented
//Constraint(validatedBy = ReCaptchaValidator.class)
Target(TYPE)
Retention(RetentionPolicy.RUNTIME)
public annotation class ReCaptcha(
        val challengeField: String = "recaptcha_challenge_field",
        val responseField: String = "recaptcha_response_field",
        val message: String = "{waypalm.site.validation.ReCaptcha.message}",
        val groups: Array<Class<*>>= array(),
        val payload: Array<Class<Payload>> = array()
)

public class ReCaptchaValidator [Inject](
        val captchaValidator: RecaptchaValidator
)
: ConstraintValidator<ReCaptcha, Any>
{
    var challengeField: String = ""
    var responseField: String = ""

    public override fun initialize(constraintAnnotation: ReCaptcha?) {
        challengeField = constraintAnnotation.challengeField();
        responseField = constraintAnnotation.responseField();
    }

    public override fun isValid(value: Any?, context: ConstraintValidatorContext?): Boolean {
        throw UnsupportedOperationException()
    }
}